include/uxr/agent/middleware/fast/FastEntities.hpp

```diff
@@ -0,0 +1,225 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef UXR_AGENT_MIDDLEWARE_FAST_FAST_ENTITIES_HPP_
+#define UXR_AGENT_MIDDLEWARE_FAST_FAST_ENTITIES_HPP_
+
+#include <fastrtps/participant/ParticipantListener.h>
+#include <fastrtps/publisher/PublisherListener.h>
+#include <fastrtps/subscriber/SubscriberListener.h>
+#include <uxr/agent/types/TopicPubSubType.hpp>
+#include <uxr/agent/middleware/Middleware.hpp>
+
+#include <unordered_map>
+#include <condition_variable>
+
+namespace eprosima {
+namespace fastrtps {
+
+class ParticipantAttributes;
+class PublisherAttributes;
+class SubscriberAttributes;
+
+} // namespace fastrtps
+} // namespace eprosima
+
+namespace eprosima {
+namespace uxr {
+
+class FastTopic;
+
+/**********************************************************************************************************************
+ * FastParticipant
+ **********************************************************************************************************************/
+class FastParticipant : public fastrtps::ParticipantListener
+{
+public:
+    FastParticipant(int16_t domain_id)
+        : domain_id_(domain_id)
+        , ptr_(nullptr)
+        , topics_register_{}
+    {}
+
+    ~FastParticipant() override;
+
+    bool create_by_ref(const std::string& ref);
+
+    // TODO (julian: #4372): add const qualifier in attrs.
+    bool create_by_attributes(fastrtps::ParticipantAttributes& attrs);
+
+    bool match_from_ref(const std::string& ref) const;
+
+    bool match_from_xml(const std::string& xml) const;
+
+    bool remove();
+
+    void onParticipantDiscovery(
+            fastrtps::Participant*,
+            fastrtps::rtps::ParticipantDiscoveryInfo&& info) override;
+
+    fastrtps::Participant* get_ptr() const { return ptr_; }
+
+    // TODO (Julian): change FastRTPS in order to accept a `const FastTopic*` ar argiment
+    bool register_topic(FastTopic* const topic,
```

Comment:
It is just for safety, that is, to avoid things like these:
```c
void silly_func(int* a)
{
    int b = 0;
    a = &b;
    *a = 1;
} 
```

Do you prefer not to use it?

Reply:
But, with the parameter you passed, you can do also the same in this way:


```c
void silly_func(int* const a)
{
    *a = 1; //this is ok
} 
```

if you write `a = &b;` by error is because you probably don't know what are you doing, as if you multiply by 4 when you want to multiply by 2. It is more a conceptual error than a misplacing error.

Personally, I think that this `const` does not protect you in any way, and obfuscate the code. I needed to stop and think about what was happening with this pointer (and nothing weird was happening!).

This problem does not occur in Rust... :heart:

Reply:
I agree with you, but would all conceptual errors might be checked by the compiler ...